db/backend_test.go

```diff
@@ -55,27 +62,89 @@ func withDB(t *testing.T, creator dbCreator, fn func(DB)) {
 }
 
 func TestBackendsNilKeys(t *testing.T) {
-	// test all backends
+	// test all backends.
+	// nil keys are treated as the empty key for most operations.
 	for dbType, creator := range backends {
 		withDB(t, creator, func(db DB) {
-			panicMsg := "expecting %s.%s to panic"
-			assert.Panics(t, func() { db.Get(nil) }, panicMsg, dbType, "get")
-			assert.Panics(t, func() { db.Has(nil) }, panicMsg, dbType, "has")
-			assert.Panics(t, func() { db.Set(nil, []byte("abc")) }, panicMsg, dbType, "set")
-			assert.Panics(t, func() { db.SetSync(nil, []byte("abc")) }, panicMsg, dbType, "setsync")
-			assert.Panics(t, func() { db.Delete(nil) }, panicMsg, dbType, "delete")
-			assert.Panics(t, func() { db.DeleteSync(nil) }, panicMsg, dbType, "deletesync")
+			t.Run(fmt.Sprintf("Testing %s", dbType), func(t *testing.T) {
+
+				expect := func(key, value []byte) {
+					if len(key) == 0 { // nil or empty
+						assert.Equal(t, db.Get(nil), db.Get([]byte("")))
+						assert.Equal(t, db.Has(nil), db.Has([]byte("")))
+					}
+					assert.Equal(t, db.Get(key), value)
+					assert.Equal(t, db.Has(key), value != nil)
+				}
+
+				// Not set
+				expect(nil, nil)
+
+				// Set nil value
+				db.Set(nil, nil)
+				expect(nil, []byte(""))
+
+				// Set empty value
+				db.Set(nil, []byte(""))
+				expect(nil, []byte(""))
+
+				// Set nil, Delete nil
+				db.Set(nil, []byte("abc"))
+				expect(nil, []byte("abc"))
+				db.Delete(nil)
+				expect(nil, nil)
+
+				// Set nil, Delete empty
+				db.Set(nil, []byte("abc"))
+				expect(nil, []byte("abc"))
+				db.Delete([]byte(""))
+				expect(nil, nil)
+
+				// Set empty, Delete nil
+				db.Set([]byte(""), []byte("abc"))
+				expect(nil, []byte("abc"))
+				db.Delete(nil)
+				expect(nil, nil)
+
+				// Set empty, Delete empty
+				db.Set([]byte(""), []byte("abc"))
+				expect(nil, []byte("abc"))
```

Comment:
minor: should this be `expect([]byte(""), []byte("abc"))`. i understand that nil and []byte("") are interchangeble

Reply:
that's why it should be nil, cuz it's easier to type :P